spec/index.adoc

```diff
@@ -2301,8 +2301,9 @@ UpdateFunc = WasmState -> Trap | Return {
   new_certified_data : NoCertifiedData | Blob
   response : NoResponse | Response;
   cycles_accepted : Nat;
+  cycles_used : Nat;
 }
-QueryFunc = WasmState -> Trap | Return Response
+QueryFunc = WasmState -> Trap | Return (Response, Nat)
```

Comment:
Such a plain `Nat` makes it hard to read, and these types may be the first someone looks at.

Maybe introduce records for the result of running a query and for the heartbeat, with a named field?

Reply:
I intend to introduce a type synonym Cycles for Nats that represent cycles. Would it help or should the response still be a record in your opinion?

Reply:
cycles is better, as it indicates the unit, but not the semantics. At least `CyclesConsumed`?

Also, don't trapping computations also consume cycles?

Reply:
I wanted to avoid switching between various type synonyms (e.g., CyclesConsumed, CyclesRefunded etc.) because type conversions between type synonyms are not applicable and type (synonym) changes without type conversions could be confusing, too. What do you think?

Trapping computations consume `if Is_response then MAX_CYCLES_PER_RESPONSE else MAX_CYCLES_PER_MESSAGE` cycles because they're not refunded anything and thus there is no need explicitly track the amount of cycles consumed.

Reply:
A trapping computation still uses `MAX_CYCLES_PER_MESSAGE`? I thought this was fixed at some point… programmer expect trapping to be a cheap way out, and if it incurs the maximum cost instead that’s not great. @bogwar?

Reply:
As for the return type `Return (Response, Nat)`, I lean towards turning it into a record `Return {response: Response, cycles_consumed: Nat}` now.

Reply:
Done.

Reply:
Even if it isn’t the case right now in the implementation, I would certainly not want the spec to forbid the implementation to do more precise cycle accounting in trapping computation. This would would suggest something like
```
 QueryFunc = WasmState -> ((Trap | Return Response), Nat) 
```
and similarly for other functions.

Reply:
> A trapping computation still uses `MAX_CYCLES_PER_MESSAGE`? I thought this was fixed at some point… programmer expect trapping to be a cheap way out, and if it incurs the maximum cost instead that’s not great. @bogwar?

it looks that we have fixed that and we refund what was not consumed

Reply:
Let me then introduce a record for cycles_used in Trap, too:
```
Trap {
  cycles_used : Nat;
}
```

Reply:
and update the Message execution transition...

Reply:
That also works of course, although I find it (on a philosophical level) less natural - trapping or returning is independent of cycle accounting. But no need to let that get in the way of a more convenient formulation :-)

Reply:
I've fixed the method returns.